tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -8,52 +8,313 @@
 #include "ProgramInfo.h"
 #include "MappingVisitor.h"
+#include <sstream>
 
 using namespace clang;
 using namespace llvm;
 
+// Helper method to print a Type in a way that can be represented in the source.
+static
+std::string
+tyToStr(const Type *T) {
+  QualType QT(T, 0);
+
+  return QT.getAsString();
+}
+
+PointerVariableConstraint::PointerVariableConstraint(DeclaratorDecl *D,
+  uint32_t &K, Constraints &CS) :
+  PointerVariableConstraint(D->getType().getTypePtr(), K, D->getName(), CS) { }
+
+PointerVariableConstraint::PointerVariableConstraint(const Type *_Ty,
+  uint32_t &K, std::string N, Constraints &CS) :
+  ConstraintVariable(ConstraintVariable::PointerVariable, tyToStr(_Ty)),
+  FV(nullptr)
+{
+  const Type *Ty = nullptr;
+  for (Ty = _Ty;
+    Ty->isPointerType();
+    Ty = getNextTy(Ty))
+  {    
+    // Allocate a new constraint variable for this level of pointer.
+    vars.insert(K);
+    CS.getOrCreateVar(K);
+    K++;
+  }
+
+  // If, after boiling off the pointer-ness from this type, we hit a 
+  // function, then create a base-level FVConstraint that we carry 
+  // around too.
+  if (Ty->isFunctionType()) {
+    FV = new FVConstraint(Ty, K, N, CS);
+  }
+
+  BaseType = tyToStr(Ty);
+
+  // Special case for void to not make _Ptr<void> pointers.
+  if( BaseType == "void" ) 
+    for (const auto &V : vars)
+      CS.addConstraint(CS.createEq(CS.getOrCreateVar(V), CS.getWild()));
+  
+}
+
+void PointerVariableConstraint::print(raw_ostream &O) const {
+  O << "{ ";
+  for (const auto &I : vars) 
+    O << "q_" << I << " ";
+  O << " }";
+}
+
+// Mesh resolved constraints with the PointerVariableConstraints set of 
+// variables and potentially nested function pointer declaration. Produces a 
+// string that can be replaced in the source code.
+std::string
+PointerVariableConstraint::mkString(Constraints::EnvironmentMap &E) {
+  // TODO: the use of 's' is inefficient, use a stringstream later.
+  std::string s = "";
+  unsigned caratsToAdd = 0;
+  bool emittedBase = false;
+  bool wroteFPtr = false;
+  for (const auto &V : vars) {
+    VarAtom VA(V);
+    ConstAtom *C = E[&VA];
+    assert(C != nullptr);
+
+    switch (C->getKind()) {
+    case Atom::A_Ptr:
+      emittedBase = false;
+      s = s + "_Ptr<";
+
+      caratsToAdd++;
+      break;
+    case Atom::A_Arr:
+    case Atom::A_Wild:
+      if (emittedBase) {
+        s = s + "*";
+      } else {
+        assert(BaseType.size() > 0);
+        emittedBase = true;
+        if (FV) {
+          s = s + FV->mkString(E);
+          wroteFPtr = true;
+        } else {
+          s = s + BaseType + "*";
+        }
+      }
+      break;
+    case Atom::A_Const:
+    case Atom::A_Var:
+      llvm_unreachable("impossible");
+      break;
+    }
+  }
+
+  if(emittedBase == false) {
+    // If we have a FV pointer, then our "base" type is a function pointer
+    // type.
+    if (FV) {
+      s = s + FV->mkString(E);
```

Comment:
I think this will generate an incorrect string in the case where the enclosing pointer type is an arr or wild pointer type.    The C syntax for pointers to function is baroque.  Given a pointer to a function that returns an integer and takes a char as an input, I think you need to write int (*)(char c);